File: internal/api/spa.go
package api

import (
	"io/fs"
	"net/http"
	"strings"
)

// SPAHandler serves an embedded filesystem with fallback to index.html for
// client-side routing. API routes should be registered before this catch-all.
func SPAHandler(fsys fs.FS) http.HandlerFunc {
	fileServer := http.FileServer(http.FS(fsys))

	return func(w http.ResponseWriter, r *http.Request) {
		// Try to serve the exact file
		path := strings.TrimPrefix(r.URL.Path, "/")
		if path == "" {
			path = "index.html"
		}

		// Check if the file exists in the embedded FS
		f, err := fsys.Open(path)
		if err == nil {
			f.Close()
			fileServer.ServeHTTP(w, r)
			return
		}

		// Fallback to index.html for client-side routing
		r.URL.Path = "/"
		fileServer.ServeHTTP(w, r)
	}
}

File: backend/src/api/spa.ts
/**
 * OpenHive Backend - SPA Handler
 *
 * Serves the compiled React SPA from disk. Static assets that exist on disk
 * are served directly; any unmatched path that is not an API route falls back
 * to index.html for client-side routing (React Router deep links).
 *
 */

import fastifyStatic from '@fastify/static';
import type { FastifyInstance } from 'fastify';

// ---------------------------------------------------------------------------
// SPAPluginOptions
// ---------------------------------------------------------------------------

export interface SPAPluginOptions {
  /** Absolute path to the compiled React SPA dist directory. */
  root: string;
}

// ---------------------------------------------------------------------------
// spaPlugin
// ---------------------------------------------------------------------------

/**
 * Fastify plugin that serves the React SPA.
 * Register all API routes BEFORE this plugin — they take priority.
 * Static files are served when they exist; all other non-API paths fall back
 * to index.html (client-side routing).
 *
 */
export async function spaPlugin(
  fastify: FastifyInstance,
  opts: SPAPluginOptions,
): Promise<void> {
  // wildcard:false → only files that exist get routes; unknown paths fall
  // through to notFoundHandler instead of being handled internally.
  await fastify.register(fastifyStatic, {
    root: opts.root,
    wildcard: false,
    index: false,
  });

  // Client-side routing fallback.
  // API paths that don't match a registered route return 404 JSON, not the SPA.
  fastify.setNotFoundHandler((request, reply) => {
    if (request.url.startsWith('/api/')) {
      return reply.status(404).send({ error: { code: 'NOT_FOUND', message: 'Not found' } });
    }
    return reply.sendFile('index.html');
  });
}
